Lock optimizations for DistAutogradContainer.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36529 Lock optimizations for DistAutogradContainer.**

DistAutogradContainer is a singleton for the entire process and has a
single lock that protects access to a map keyed by `context_id`. Performance
profiling showed that this lock is a potential bottleneck for training. As a
result, in this PR we have the following optimizations:

1) Shard the map into multiple buckets with each bucket having its own lock. This
would ensure we hold much finer grained locks.
2) sendReleaseContextRpc was being called under a lock, moved this to be
outside the lock.

Differential Revision: [D21003934](https://our.internmc.facebook.com/intern/diff/D21003934/)

pritamdamania committed:
Lock optimizations for DistAutogradContainer.

DistAutogradContainer is a singleton for the entire process and has a
single lock that protects access to map keyed by context_id. Performance
profiling showed that this lock is a potential bottleneck for training. As a
result, in this PR, we have the following optimizations:

1) Shard the map into 256 buckets with each bucket having its own lock. This
would ensure we hold much finer grained locks.
2) sendReleaseContextRpc was being called under a lock, moved this to be
outside the lock.

Differential Revision: [D21003934](https://our.internmc.facebook.com/intern/diff/D21003934/)

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/autograd/context/container.h:
```diff
@@ -91,6 +91,18 @@ class TORCH_API DistAutogradContainer {
+    std::unordered_map<int64_t, ContextPtr> contexts;
+
+    // Lock for this shard.
+    mutable std::mutex lock;
```
- consider reordering so that lock is first (acquiring the lock will fetch the beginning part of the unordered_map<>  into cache in exclusive mode automatically)

- consider padding slightly, to avoid interference.
    in fbcode, sizeof() this struct should be 96 bytes. If the hardware cache line is 128, two adjacent entries will contend with each other.
   std::hardware_destructive_interference_size is c++17-only, but could always just add something like the following to the bottom, and call it a day:
       int64_t unusedCachePadding_[8];  // prevent adjacent shards from sharing a cache line

jjlilley replied:
oh, so I'm learning that the more modern way is something like

```
 struct alignas(128) ContextsShard {
```

wanchaol commented on torch/csrc/distributed/autograd/context/container.cpp:
```diff
@@ -98,20 +107,23 @@ const ContextPtr DistAutogradContainer::newContext() {
 
-  std::lock_guard<std::mutex> guard(autograd_context_lock_);
+  auto context_id = next_context_id_++;
+  current_context_id_ = context_id;
```
what's the difference between `context_id` and `current_context_id`?

wanchaol commented on torch/csrc/distributed/autograd/context/container.h:
```diff
@@ -91,6 +91,18 @@ class TORCH_API DistAutogradContainer {
+  };
+
+  // Number of shards for the map storing autograd contexts.
+  static constexpr int16_t kNumShards = 256;
```
should we use some heuristic formula (sth like @jjlilley mentioned above) for the default numShards?

jjlilley replied:
btw, I wasn't necessarily advocating setting size based on num_cpus... (I kind of like the power-of-2 mode here)

Just saying (without data), that my guess is that the perf gain from 128 to 256 might be low

pritamdamania committed:
Update on "Lock optimizations for DistAutogradContainer."


DistAutogradContainer is a singleton for the entire process and has a
single lock that protects access to a map keyed by `context_id`. Performance
profiling showed that this lock is a potential bottleneck for training. As a
result, in this PR we have the following optimizations:

1) Shard the map into 256 buckets with each bucket having its own lock. This
would ensure we hold much finer grained locks.
2) sendReleaseContextRpc was being called under a lock, moved this to be
outside the lock.

Differential Revision: [D21003934](https://our.internmc.facebook.com/intern/diff/D21003934/)

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/autograd/context/container.cpp:
```diff
@@ -50,6 +56,29 @@ DistAutogradContainer& DistAutogradContainer::init(int64_t worker_id) {
+      num_shards <<= 1;
+    }
+  }
+  LOG(INFO) << "Number of shards for DistAutogradContainer: " << num_shards;
```
This log is fine for now, though eventually we might want to demote to VLOG(1)
